monai/auto3dseg/data_analyzer.py

```diff
@@ -0,0 +1,143 @@
+            datalist to locate the label files of the dataset. If label_key is None, the DataAnalyzer
+            will skip looking for labels and all label-related operations.
+
+    For example:
+
+    .. code-block:: python
+
+        from monai.apps.auto3dseg.data_analyzer import DataAnalyzer
+
+        datalist = {
+            "testing": [{"image": "image_003.nii.gz"}],
+            "training": [
+                {"fold": 0, "image": "image_001.nii.gz", "label": "label_001.nii.gz"},
+                {"fold": 0, "image": "image_002.nii.gz", "label": "label_002.nii.gz"},
+                {"fold": 1, "image": "image_001.nii.gz", "label": "label_001.nii.gz"},
+                {"fold": 1, "image": "image_004.nii.gz", "label": "label_004.nii.gz"},
+            ],
+        }
+
+        dataroot = '/datasets' # the directory where you have the image files (nii.gz)
+        DataAnalyzer(datalist, dataroot)
+
+    Notes:
+        The module can also be called from the command line interface (CLI).
+
+    For example:
+
+    .. code-block:: bash
+
+        python -m monai.apps.auto3dseg \
+            DataAnalyzer \
+            get_all_case_stats \
+            --datalist="my_datalist.json" \
+            --dataroot="my_dataroot_dir"
+
+    """
+
+    def __init__(
+        self,
+        datalist: Union[str, Dict],
+        dataroot: str = "",
+        output_path: str = "./data_stats.yaml",
+        do_ccp: bool = True,
+        device: Union[str, torch.device] = "cuda",
+        worker: int = 2,
+        image_key: str = "image",
+        label_key: str = "label",
+    ):
+        """
+        The initializer will load the data and register the functions for data statistics gathering.
+        """
+        if path.isfile(output_path):
+            warnings.warn(f"File {output_path} already exists and will be overwritten.")
+            logger.debug(f"{output_path} will be overwritten by a new datastat.")
+
+        self.image_key = image_key
+        self.label_key = label_key
+
+        self.output_path = output_path
+        self.IMAGE_ONLY = True if label_key is None else False
+
+        self.datalist = datalist
+        self.dataroot = dataroot
+        self.device = device
+        self.worker = worker
+
+    def get_all_case_stats(self):
+
+        keys = list(filter(None, [self.image_key, self.label_key]))
+        files, _ = datafold_read(datalist=self.datalist, basedir=self.dataroot, fold=-1)
+        ds = data.Dataset(data=files)
+        self.dataset = data.DataLoader(
+            ds, batch_size=1, shuffle=False, num_workers=self.worker, collate_fn=no_collation
+        )
+
+        result = {DATA_STATS.SUMMARY: {}, DATA_STATS.BY_CASE: []}
+
+        case_engine = SegAnalyzeCaseEngine(self.image_key, self.label_key, device=self.device)
+        for batch_data in self.dataset:
+            result[DATA_STATS.BY_CASE].append(case_engine(batch_data[0]))
```

Comment:
The case engine `case_engine(batch_data[0])` still looks like a MapTransform call, where `batch_data` is a batch size 1 item, could we make it a transform? The developers would learn one less new concept of `CaseEngine`. For example, the usage would be:

```py
case_engine = SegAnalyzeCaseEngine(self.image_key, self.label_key, device=self.device)
case_engine.update({"my_stats_op": np.nanstd})
case_engine.update({"image_foreground_stats": FgImageStatsCasesAnalyzer(image_key, label_key)})

transform_list = [
    LoadImaged(keys=keys),
    EnsureChannelFirstd(keys=keys),  # this creates label to be (1,H,W,D)
    ToDeviced(keys=keys, device=device, non_blocking=True),
    Orientationd(keys=keys, axcodes="RAS"),
    EnsureTyped(keys=keys, data_type="tensor"),
    Lambdad(keys=label_key, func=lambda x: torch.argmax(x, dim=0, keepdim=True) if x.shape[0] > 1 else x)
    if label_key
    else None,
    SqueezeDimd(keys=["label"], dim=0) if label_key else None,
    case_engine,
]

ds = data.Dataset(data=files, transform=Compose(transform_list))
for case_result in ds:
    result[BY_CASE].append(case_result)
```



the `SegAnalyzeCaseEngine` class could be something like:
```py
class SegAnalyzeCaseEngine(MapTransform):
    def __init__(self) -> None:
        self.analyzers = {}

    def update(self, analyzer: Dict[str, callable]):
        self.analyzers.update(analyzer)

    def __call__(self, data):
        ret = {}
        for k, analyzer in self.analyzers.items():
            if callable(analyzer):
                ret.update({k: analyzer(data)})
            elif isinstance(analyzer, str):
                ret.update({k: analyzer})
        return ret
```

in this way it's possible to run the per case study in parallel, even across nodes https://github.com/Project-MONAI/tutorials/blob/main/modules/compute_metric.py if we  use an aggregator.

I think this `DataAnalyzer` API should at least give some flexibility of leveraging the fact that the per-case stats are independent of each other and could be computed in parallel.

Reply:
@wyli @Nic-Ma 

As a follow up, we can do the following after commit 3e5c65587593f3f1897930f52bec209aaaab695d. `SegAnalyzeEngine` now inherits `Compose` with the following usage:

```python
analyze_engine = SegAnalyzeEngine("image", "label")
transform_list = [
    ...
    Lambdad(keys="label", func=lambda x: torch.argmax(x, dim=0, keepdim=True) if x.shape[0] > 1 else x),
    SqueezeDimd(keys=["label"], dim=0),
    analyze_engine,
]
...
transform = Compose(transform_list)
stats = []
for batch_data in self.dataset:
    d = transform(batch_data[0])
    stats.append(d)
report = analyze_engine.summarize(stats)
```

With the new SegAnalyzeEngine, developers can add new analyzers by
```python
analyze_engine = SegAnalyzeEngine("image", "label")
user_case_analyzer = ...
user_summary_analzer = ...   # this can be None if summary is not needed

analyze_engine.add_analyzer(user_case_analyzer, user_summary_analzer)

```

Reply:
thanks, this looks nice. (for the class names, how about renaming `SegAnalyzeEngine` to `SegSummarizer` and the analyzers such as `FgImageStatsCaseAnalyzer` to `FgImageStats`.)